flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,224 @@
+// then you can get the jobmastergateway (the actual gateway)
+
+// happy case: request ArchivedExecutionGraph from JM on success, otherwise, FailedArchiveExecGraph
+// AutoCloseableAsync call Dispatcher.removeJob()
+// CompletableFuture<ArchivedExecutionGraph> getResultFuture();
+
+// -------------------- new DispatcherJob -------------------------- //
+
+/**
+ * TODO.
+ */
+public class DispatcherJob implements AutoCloseableAsync {
+
+	protected final Logger log = LoggerFactory.getLogger(getClass());
+
+	private final CompletableFuture<JobManagerRunner> jobManagerRunnerFuture;
+	private final CompletableFuture<ArchivedExecutionGraph> jobResultFuture;
+	// job is running when this completes
+	private final CompletableFuture<JobMasterGateway> jobMasterGatewayFuture;
+
+	private final JobGraph jobGraph;
+	private final long initializationTimestamp;
+
+	// if this future is != null, the job has been cancelled. Cancellation completes with this future.
+	@Nullable
+	private CompletableFuture<Acknowledge> cancellationFuture = null;
+
+	static DispatcherJob createForSubmission(CompletableFuture<JobManagerRunner> jobManagerRunnerFuture, JobGraph jobGraph, long initializationTimestamp) {
+		return new DispatcherJob(jobManagerRunnerFuture, jobGraph, initializationTimestamp);
+	}
+
+	static DispatcherJob createForRecovery(CompletableFuture<JobManagerRunner> bla) {
+		return null; // TODO handle HA case
+	}
+
+	private DispatcherJob(CompletableFuture<JobManagerRunner> jobManagerRunnerFuture,
+		JobGraph jobGraph, long initializationTimestamp) {
+		this.jobManagerRunnerFuture = jobManagerRunnerFuture;
+		this.jobGraph = jobGraph;
+		this.initializationTimestamp = initializationTimestamp;
+		jobResultFuture = new CompletableFuture<>();
+		jobMasterGatewayFuture = new CompletableFuture<>();
+		FutureUtils.assertNoException(jobManagerRunnerFuture.handle((jobManagerRunner, throwable) -> {
+			// this gets called when the JM has been initialized
+			log.info("jm runner handle called", throwable);
+			if (throwable == null) {
+				if (cancellationFuture != null) {
+					log.warn("JobManager initialization has been cancelled for {}. Stopping JobManager.", jobGraph.getJobID());
+					// todo consider exposing a method in the JobManagerRunner to properly forward an exception
+					initializationFailedWith(null, JobStatus.CANCELED);
+					CompletableFuture<Void> jmCloseFuture = jobManagerRunner.closeAsync();
+					FutureUtils.assertNoException(jmCloseFuture.whenComplete((ign, ore) -> {
+						cancellationFuture.complete(Acknowledge.get());
+					}));
+					log.warn("cancellation future right now: " + cancellationFuture);
+					cancellationFuture.handleAsync((ack, thr) -> {
+						log.info("cancellation future finished with " + ack + " thr = " + thr);
+						return null;
+					});
+				} else {
+					// request JobMaster gateway
+					jobManagerRunner
+						.getJobMasterGateway()
+						.thenAccept(jobMasterGatewayFuture::complete);
```

Comment:
I think we need to handle changing leaderships of the `JobMaster`. `jobManagerRunner.getJobMasterGateway()` might return an outdated `JobMasterGateway` belonging to a previous leader session.